src/metabase/public_settings/premium_features.clj

```diff
@@ -240,3 +241,169 @@
+  (let [[schema-indicator [schema & more]] (u/optional #{:-} args)]
+    (if schema-indicator
+      [schema more]
+      [nil args])))
+
+(defn- parse-defenterprise-args
+  [defenterprise-args]
+  (let [[result-schema more]    (parse-result-schema defenterprise-args)
+        [docstr more]           (u/optional string? more)
+        [ee-ns more]            (u/optional symbol? more)
+        [options [args & more]] (parse-defenterprise-options more)
+        [arg->schema body]      (u/optional (every-pred map? #(every? symbol? (keys %))) more)]
+    {:ee-ns         ee-ns
+     :docstr        docstr
+     :options       options
+     :args          args
+     :arg->schema   arg->schema
+     :result-schema result-schema
+     :body          body}))
+
+(defn validate-args
+  "Generate a series of s/validate calls for the arg and schema pairs in `arg->schema`."
+  [arg->schema]
+  (for [[arg schema] arg->schema]
+    `(s/validate ~schema ~arg)))
+
+(defonce ^{:doc "A map from EE functions (as [namespace function] tuples) to anonymous fns which have the same body as
+                their OSS equivalents. These fns are called when the EE function must fallback to the OSS behavior due
+                to absence of a feature flag."}
+  ee-registry
+  (atom {}))
+
+(defn- register-mapping!
+  "Adds mapping from EE->OSS fn to the `ee-registry`."
+  [fn-name ee-ns args body]
+  (let [ee-parts  [ee-ns fn-name]
+        oss-fn    (eval `(fn ~args ~@body))]
+    (swap! ee-registry assoc ee-parts oss-fn)))
+
+(defn missing-premium-token-exception
+  "The default exception to throw when an EE function is called, but a required premium feature is not present."
+  [fn-name feature]
+  (let [message (if (= feature :any)
+                  (trs "The {0} function requires a valid premium token"
+                       fn-name)
+                  (trs "The {0} function requires a valid premium token with the {1} feature"
+                       fn-name
+                       (name feature)))]
+    (ex-info message {:status-code 402})))
+
+(defmacro defenterprise-ee
+  "Impl macro for `defenterprise` when used in an EE namespace. Don't use this directly."
+  [{:keys [fn-name docstr args body result-schema options]}]
+  (let [{:keys [feature fallback]} options]
+    `(if ~(or (= feature :none)
+              (if (or (not feature) (= feature :any))
+                `(enable-enhancements?)
+                `(has-feature? ~feature)))
+
+       ~(if result-schema
+          `(s/validate ~result-schema (do ~@body))
+          `(do ~@body))
+
+       ~(cond
+          (= fallback :error)
+          `(throw (missing-premium-token-exception ~(str fn-name) ~feature))
+
+          (or (symbol? fallback) (seq? fallback))
+          `(apply ~fallback ~args)
+
+          ;; :oss and default case
+          :else
+          `(apply (get @ee-registry [(symbol ~(str (ns-name *ns*)))
+                                     (symbol ~(str fn-name))])
+                  ~args)))))
```

Comment:
If I were you I'd do all this feature check stuff inside of a function rather than in the every macroexpansion itself. It's way easier to write fancy macros if you try to do as much stuff outside of the macro as you can.

Use the macro for syntax transformation and have it hand off to plain functions as much as possible